src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Automation/Peers/GridViewAutomationPeer.cs

```diff
@@ -195,7 +195,26 @@ int IGridProvider.ColumnCount
         /// </summary>
         int IGridProvider.RowCount
         {
-            get { return _listview.Items.Count; }
+            get 
+            { 
+                int rowcount = _listview.Items.Count;
+                foreach(var item in _listview.Items)
+                {
+                    if (item is UIElement { Visibility: not Visibility.Visible })
+                    {
+                        rowcount -= 1;
+                        continue;
+                    }
+        
+                    var container = _listview.ItemContainerGenerator.ContainerFromItem(item);
```

Comment:
Doesn't this return the item if it is the container? Also the item being UIElement does not necessarily mean it is part of the tree.

Reply:
For the record this was is the example showing that checking visibility on items is misleading:

```C#
<ListView>
    <ListView.View>
        <GridView>
            <GridViewColumn Header="Visibility">
                <GridViewColumn.CellTemplate>
                    <DataTemplate>
                        <TextBlock Text="{Binding Visibility}" />
                    </DataTemplate>
                </GridViewColumn.CellTemplate>
            </GridViewColumn>
        </GridView>
    </ListView.View>
    <ListView.Items>
        <TextBlock Visibility="Visible" Text="TB visible" />
        <TextBlock Visibility="Hidden" Text="TB hidden" />
        <TextBlock Visibility="Collapsed" Text="TB collapsed" />
        <TextBlock Visibility="Visible" Text="TB visible" />
    </ListView.Items>
</ListView>
```

This will show all 4 items despite them being UIElements with not Visible Visibility.

Reply:
> Doesn't this return the item if it is the container?

That was a performance concern where you would be checking an element twice. However, the first check of an item directly should simply be dropped, so it's fine.

Reply:
> For the record this is an example showing that checking visibility on items is misleading:
> 
> ```
> <ListView>
>     <ListView.View>
>         <GridView>
>             <GridViewColumn Header="Visibility">
>                 <GridViewColumn.CellTemplate>
>                     <DataTemplate>
>                         <TextBlock Text="{Binding Visibility}" />
>                     </DataTemplate>
>                 </GridViewColumn.CellTemplate>
>             </GridViewColumn>
>         </GridView>
>     </ListView.View>
>     <ListView.Items>
>         <TextBlock Visibility="Visible" Text="TB visible" />
>         <TextBlock Visibility="Hidden" Text="TB hidden" />
>         <TextBlock Visibility="Collapsed" Text="TB collapsed" />
>         <TextBlock Visibility="Visible" Text="TB visible" />
>     </ListView.Items>
> </ListView>
> ```
> 
> This will show all 4 items despite them being UIElements with not Visible Visibility.

Yes, this is a valid case and our current fix won't work. It will give an incorrect result.
Thanks for bringing this up.